src/opentracing-c/propagation.h

```diff
@@ -0,0 +1,220 @@
+
+    /**
+     * Occurs when inject() or extract() implementations expect a different type
+     * of carrier than they are given.
+     */
+    opentracing_propagation_error_code_invalid_carrier = -4,
+
+    /**
+     * Occurs when the carrier passed to extract() is of the expected type but
+     * is corrupted.
+     */
+    opentracing_propagation_error_code_span_context_corrupted = -5
+} opentracing_propagation_error_code;
+
+/**
+ * Propagation formats.
+ */
+typedef enum opentracing_propagation_format {
+    /** Represents span contexts as opaque binary data. */
+    opentracing_propagation_format_binary = 1,
+
+    /**
+     * Represents span contexts as key:value string pairs. Unlike
+     * opentracing_propagation_format_http_headers, the
+     * opentracing_propagation_format_text_map format does not restrict
+     * the key or value character sets in any way.
+     */
+    opentracing_propagation_format_text_map,
+
+    /**
+     * Represents span contexts as HTTP header string pairs. Unlike
+     * opentracing_propagation_format_text_map, the
+     * opentracing_propagation_format_http_headers requires that the
+     * keys and values be valid as HTTP headers as-is (i.e., character casing
+     * may be unstable and special characters are disallowed in keys, values
+     * should be URL-escaped, etc.).
+     */
+    opentracing_propagation_format_http_headers,
+
+    /**
+     * Represents span contexts in a user-defined format. With it, the caller
+     * can encode/decode a opentracing_span_context for propagation as entries
+     * in a custom protocol.
+     */
+    opentracing_propagation_format_custom
+} opentracing_propagation_format;
+
+/**
+ * The inject carrier for the opentracing_propagation_format_text_map.
+ * With it, the caller can encode a span context for propagation as entries in a
+ * map of strings.
+ * @attention The backing store for the opentracing_text_map_writer may
+ *            contain data unrelated to span context. As such, inject() and
+ *            extract() implementations that call the
+ *            opentracing_text_map_writer and opentracing_text_map_reader
+ *            interfaces must agree on a prefix or other convention to
+ *            distinguish their own key:value pairs.
+ * @extends opentracing_destructible
+ */
+typedef struct opentracing_text_map_writer {
+    /** Base class member. */
+    opentracing_destructible base;
+
+    /**
+     * Set a key:value pair to the carrier. Multiple calls to set() for the
+     * same key leads to undefined behavior.
+     * @param writer Writer instance.
+     * @param key String key.
+     * @param value String value.
+     * @return opentracing_propagation_error_code indicating success or failure.
+     */
+    opentracing_propagation_error_code (*set)(
+        struct opentracing_text_map_writer* writer,
+        const char* key,
+        const char* value) OPENTRACINGC_NONNULL(1);
```

Comment:
What gets returned if there's a non-propagation error (e.g. memory failure)?

Reply:
Good question. Would you suggest I use an int return type instead or add a new generic failure code?

Reply:
`int` along with an unknown error code could be one way to solve it. (Looks like that's what mysql's c api does: https://dev.mysql.com/doc/refman/5.7/en/mysql-query.html).

Reply:
I usually see `int` coupled with tons of macros (`errno.h` is a good example), or pure enums. I'm trying to keep the enum for now. Will see if that makes sense.